### DISCUSSION TITLE
📸 Gallery: Post your screenshots here

### DISCUSSION THREAD

**Discussion by brenocq** [brenocq/implot3d#3] [Gallery]
Welcome to the ImPlot3D Gallery 📸! This is the place to share screenshots, GIFs, or short videos of your 3D visualizations!

I would love to see how you're using ImPlot3D in your application! ❤️

**Comment by brenocq** [comment]
Amazing to see [BrutPitt's post on X](https://x.com/BrutPitt/status/1869179122838286682) using ImPlot3D to visualize the **Thomas Attractor**. As far as I know, this was the first use of ImPlot3D shared online (excluding me). Super exciting, I can't wait to see ImPlot3D being used in other applications 😍 

https://github.com/user-attachments/assets/01a171e4-02ba-4d49-adb7-db2f7ae13f7f

**Reply by dcnieho** [reply]
ImPlot3D is included in https://github.com/pthom/imgui_bundle, which makes me very excited as now i'm able to use it. thanks for your work!

**Reply by brenocq** [reply]
Awesooome!! hehe let me know if you have any problems with ImPlot3D or if you would like new types of plot to be implemented :)

**Reply by dcnieho** [reply]
Well, wasn't me who did that, all credit to @pthom. I can't oversell how happy i am with his library. Thank you too man, this is something I was just starting to realize i needed it, only to find you have created it :)

**Reply by brenocq** [reply]
Thanks for the kind words! @pthom did indeed some insane work kkkkkk he even implemented a super cool Lorenz Attractor in #45

**Comment by fecaltransplant** [comment]
A much-needed addition to Dear ImGui, and fantastically implemented!  I only have praise and the slightest of nitpicks.
 
I'm plotting histograms over time.  This looks like it'll work as is as long as I throw out some measurements (this is every 8th histogram from the data set) due to the limits set by ImDrawList3D::MaxIdx() if I understand correctly.  The discussion about making adjustments based on level of detail sounds fun.  In this dataset, fully 50 percent of the histogram is flat.  If such sorcery could be implemented to where I could throw in the whole data set and have its rendering optimized on the fly, I'd be delighted.

These are log distributions, so a log10 axis format will be greatly appreciated if it isn't there already.  _I think this is really the only thing I truly lack at this time._

It already looks like (based on the coordinates) that I'll be able to implement a feature to click on the graph and return the histogram selected.

I also commented out the render lines part out of PlotSurfaceEx().  Maybe a flag or checkbox to toggle that?

Also, in my data I have fixed x and y value ranges.  Would it be possible to enable that option in the plot surface to ease the entry of the x and y value ranges (without repeating them over and over in their respective arrays?)

![waterfall](https://github.com/user-attachments/assets/de9ca8cc-d551-458b-9687-f94ecf683316)

**Reply by brenocq** [reply]
Heey! Thank you a lot for the feedback ❤️ 

> I'm plotting histograms over time. This looks like it'll work as is as long as I throw out some measurements (this is every 8th histogram from the data set) due to the limits set by ImDrawList3D::MaxIdx() if I understand correctly.

Hey! Did you try to enable 32-bit vertex indices? You can plot way more vertices with it if your backend supports it. You can enable it by defining `#define ImDrawIdx unsigned int` before including ImGui. Please let me know if you experience problems with performance.

I created a feature request #111 to work on making ImPlot3D work well even with 16-bit vertex indices.

> The discussion about making adjustments based on level of detail sounds fun. In this dataset, fully 50 percent of the histogram is flat. If such sorcery could be implemented to where I could throw in the whole data set and have its rendering optimized on the fly, I'd be delighted.

That sounds awesome! I'll keep in mind to optimize both line plots and surface plots :)

> These are log distributions, so a log10 axis format will be greatly appreciated if it isn't there already. I think this is really the only thing I truly lack at this time.

Agreed! I created a feature request about it #112

> It already looks like (based on the coordinates) that I'll be able to implement a feature to click on the graph and return the histogram selected.

There are APIs to get the coordinates of your mouse on each of the three planes from the box, but they do not take your plotted data into account, would you need it to take your data into account? (like returning the 3D position on the plotted surface instead of on the box planes)

> I also commented out the render lines part out of PlotSurfaceEx(). Maybe a flag or checkbox to toggle that?

You should be able to achieve the same result by setting the line color to transparent. If the line/fill/marker color is transparent, the line/fill/marker will not be rendered. Were you expecting to use something like `ImPlot3DSurfaceFlags_NoLines` instead?

> Also, in my data I have fixed x and y value ranges. Would it be possible to enable that option in the plot surface to ease the entry of the x and y value ranges (without repeating them over and over in their respective arrays?)

I love the idea! I created a feature request about it #110

Feel free to create more discussions when you have more ideas :)

**Reply by fecaltransplant** [reply]
Thanks for the reminder on the 32-bit vertices.  I made that change in imconfig.h and it worked.  The framerate dropped but I could still use that to create a function where the user could save a "full-resolution image" versus the preview they're looking at in real-time.  

For picking points from the graph, the typical workflow will work fine looking down at the x-y plane (viewing above in the -z direction).  I love the ease of double right-clicking on the x-y plane so that I can easily isolate the x-y view.  However, I see your point that the coordinates wouldn't work as a raycast function.  I wouldn't require that, but that would be fancy.

I suppose I was looking for a flag for no lines (to remove lines) for us slower children who didn't think of transparency.

As for additional thoughts:

I notice that the box scale can be adjusted; would it be possible to snap the box to the data so that if I scale the data or move it around, that the box would scale/move along with the data?  

Would it be possible to remove/hide the box and only see the data and the axes/ticks? 

Thanks again!

**Reply by brenocq** [reply]
Sorry for the late reply @fecaltransplant! I didn't reply right away and forgot about it ;-;

> I notice that the box scale can be adjusted; would it be possible to snap the box to the data so that if I scale the data or move it around, that the box would scale/move along with the data?

That is not currently possible, but it sounds great! Maybe a flag like `ImPlot3DAxisFlags_AutoBoxScale` could be added for this. I've created the feature request #122 for this.

> Would it be possible to remove/hide the box and only see the data and the axes/ticks?

Another great idea! I think it is already possible to do it by customizing the plot style (`ImPlot3DStyle::Colors`)

![image](https://github.com/user-attachments/assets/66411d93-bf18-4c34-8504-66f99e604ad5)

Let me know if modifying the `ImPlot3DStyle` didn't work.
